File: algorithms/0028.implement_strstr/implement_strstr.rs
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        if haystack.len() < needle.len() {
            return -1;
        }
        if haystack.is_empty() && needle.is_empty() {
            return 0;
        }

        let (haystack, needle) = (haystack.as_bytes(), needle.as_bytes());
        let (mut lo, hi) = (0usize, haystack.len());
        while lo < hi {
            let mut i = 0usize;
            while lo < hi && i < needle.len() && haystack[lo] == needle[i] {
                lo += 1;
                i += 1;
            }
            // if lo >= 5 {
            //     println!(
            //         "lo : {}, i : {} haystack: {:?}, needle: {:?}",
            //         lo, i, haystack, needle
            //     )
            // }
            if i >= needle.len() {
                return (lo - i) as i32;
            }
            lo = lo - i + 1;
        }
        -1
    }
}

File: algorithms/0028.implement_strstr/implement_strstr.cpp
class Solution {
public:
    vector<int> buildNext(const string& needle) {
        vector<int> next(needle.size());
        int j = -1;
        next[0] = j;
        for (int i = 1; i < needle.size(); ++i) {
            while (j >= 0 && needle[i] != needle[j + 1]) {
                j = next[j];
            }
            if (needle[i] == needle[j + 1]) {
                ++j;
            }
            next[i] = j;
        }
        return move(next);
    }
    int strStr(const string &haystack, const string &needle) {
        if (needle.empty()) {
            return 0;
        }
        const auto next = buildNext(needle);
        int j = -1;
        for (int i = 0; i < haystack.size(); ++i) {
            while (j >= 0 && haystack[i] != needle[j + 1]) {
                j = next[j];
            }
            if (haystack[i] == needle[j + 1]) {
                ++j;
            }
            if (j == (needle.size() - 1)) {
                return i - needle.size() + 1;
            }
        }
        return -1;
    }
};